lib/rules/array-bracket-newline.js

```diff
@@ -0,0 +1,247 @@
+            category: "Stylistic Issues",
+            recommended: false
+        },
+        fixable: "whitespace",
+        schema: [
+            {
+                oneOf: [
+                    {
+                        enum: ["always", "never"]
+                    },
+                    {
+                        type: "object",
+                        properties: {
+                            multiline: {
+                                type: "boolean"
+                            },
+                            minItems: {
+                                type: "integer",
+                                minimum: 0
+                            }
+                        },
+                        additionalProperties: false
+                    }
+                ]
+            }
+        ]
+    },
+
+    create(context) {
+        const sourceCode = context.getSourceCode();
+
+
+        //----------------------------------------------------------------------
+        // Helpers
+        //----------------------------------------------------------------------
+
+        /**
+         * Normalizes a given option value.
+         *
+         * @param {string|Object|undefined} value - An option value to parse.
+         * @returns {{multiline: boolean, minItems: number}} Normalized option object.
+         */
+        function normalizeOptionValue(value) {
+            let multiline = false;
+            let minItems = Number.POSITIVE_INFINITY;
+
+            value = value || "always";
+
+            if (value) {
+                if (value === "always") {
+                    minItems = 0;
+                } else if (value === "never") {
+                    minItems = Number.POSITIVE_INFINITY;
+                } else {
+                    multiline = Boolean(value.multiline);
+                    minItems = value.minItems || Number.POSITIVE_INFINITY;
+                }
+            } else {
+                multiline = true;
+            }
+
+            return { multiline, minItems };
+        }
+
+        /**
+         * Normalizes a given option value.
+         *
+         * @param {string|Object|undefined} options - An option value to parse.
+         * @returns {{ArrayExpression: {multiline: boolean, minItems: number}, ArrayPattern: {multiline: boolean, minItems: number}}} Normalized option object.
+         */
+        function normalizeOptions(options) {
+            if (options && (options.ArrayExpression || options.ArrayPattern)) {
+                return {
+                    ArrayExpression: normalizeOptionValue(options.ArrayExpression),
+                    ArrayPattern: normalizeOptionValue(options.ArrayPattern)
+                };
+            }
+
+            const value = normalizeOptionValue(options);
+
+            return { ArrayExpression: value, ArrayPattern: value };
+        }
+
+        /**
+        * Reports that there shouldn't be a break after the first token
```

Comment:
could we change the description here and the function name to say "linebreak" instead of "break" (and in all subsequent functions below this one)? I think that would be clearer!